### PR TITLE
8260368: [PPC64] GC interface needs enhancement to support GCs with load barriers

### PR DESCRIPTION
**Pull request by Quaffel**
At present, the `needs_frame` flag is used on the ppc platform to determine whether gc barriers must emit a new stack frame (and save the link register, for that matter) or not. With the introduction of load reference barriers, however, this mechansim is no longer sufficient. This holds especially true for compiler stubs as those make heavy use of volatile registers.

To mitigate this, this patch replaces the `needs_frame` flag with a simple enumeration. As the enumerators are incremental, handling the different "register preservation needs" in the actual gc barrier implementations is comparatively (pun intended) easy. 

_This is a preparational change for the ShenandoahGC port to ppc. As such, it may provide some functionality this version doesn't make use of, but that is required for the upcoming change. This way, the scope of the upcoming change is limited to GC-specific functionality; making its review a little easier._

_For the same reason, this patch also introduces patching support for `LIR_Assembler::leal`._

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8260368](https://bugs.openjdk.java.net/browse/JDK-8260368): [PPC64] GC interface needs enhancement to support GCs with load barriers

### Reviewers
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to 46bfce1693224e014a7c46edc982e6c0541c0f3a
 * [Goetz Lindenmaier](https://openjdk.java.net/census#goetz) (@GoeLin - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2302/head:pull/2302`
`$ git checkout pull/2302`

### CODE REVIEW

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for overworking it. Looks good. I'll sponsor it after review is completed.

File: src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp [hunk 1]
```diff
@@ -108,7 +108,7 @@ void G1BarrierSetAssembler::gen_write_ref_array_post_barrier(MacroAssembler* mas
 }
 
 void G1BarrierSetAssembler::g1_write_barrier_pre(MacroAssembler* masm, DecoratorSet decorators, Register obj, RegisterOrConstant ind_or_offs, Register pre_val,
-                                                 Register tmp1, Register tmp2, bool needs_frame) {
+                                                 Register tmp1, Register tmp2, unsigned int preservation_level) {
```
Comment (rkennke) [inline src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp:111]: Wouldn't it make sense to use an enum type here? Or is this something that the PPC toolchain doesn't like? Applies to all occurances of similar change all over the place.

**Review by rkennke — COMMENTED** [review]
Thank you for doing that! I cannot really comment much on PPC specifics. Structurally and conceptually it looks reasonable to me. I've one question about the use of int vs enum though.

File: src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp (hunk 1 above)
Comment (TheRealMDoerr) [reply src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp:111]: Hi Roman,
enum type works. We only don't want to use C++11 features like enum class or typed enums because we will probably backport this change to 11u.

Comment (rkennke) [reply src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp:111]: IOW, can we write MacroAssembler::RuntimeInvocationPreservationLevel instead of unsigned int there or not? I believe this should be possible (even though it'll still only be treated by the compiler as an int)

Comment (Quaffel) [reply src/hotspot/cpu/ppc/gc/g1/g1BarrierSetAssembler_ppc.cpp:111]: Yes, that should be totally fine.

As a site note: The compiler's behavior isn't well-defined, though. It could be `unsigned int`, `int`, or any other integral type that can represent the enumerators (that is not bigger than `unsigned int`). I'm not a huge fan of _potentially_ using signed integers for enum and/or flag-like parameters, but as an enum _does_ increase the code's readability and ultimately the developer experience by a lot, we should probably use that instead. Typed enums would have been perfect here, but Martin has already pointed out why that isn't an option.

The true reason why I didn't use an enum in the first place, however, is a different one: I originally intended to implement a flag-like mechanism. As it turns out, that is an absolute overkill as the different preservation levels are incremental. Since then, I haven't touched the types, so yeah...

TL;DR: Good catch, will change it.

**Review by TheRealMDoerr — APPROVED** [review]
Even better, now. Thanks!

**Review by rkennke — APPROVED** [review]
Looks good!

**Comment by Quaffel** [comment]
> You remove a row of line breaks manking them quit long.
> In other files, you break the lines very nicely,
> with register arguments on one line, other args in
> another line etc.

My rule of thumb is: If a line fits into the (nowadays standard) 120 characters line limit quite nicely, I usually don't use additional line breaks. If it doesn't, I do what you've just described.

As far as I can tell, I'm not removing any line breaks that are necessary for keeping this limit. Nonetheless, as the variable names are very compact, the rule might not be as applicable as in other codebases. 

IMHO, it's a matter of taste. As you mention it explicitly, I'll go back to the more aggressive approach.

> interp_masm_ppc_64.cpp
> 
> Please add a comment to load_resolved_reference_at_index()
> that the index register content is destroyed.
> 
> macroAssembler_ppc.hpp:
> 
> RuntimeInvocationPreservationLevel
> Document what it is good for, not why you use what C++ construct.
> Maybe "Indicate which registers must be preserved when calling into the runtime." ?
> 

That comment was removed in dbc18a3 because it really didn't provide any value.
Suggestion looks good, I'll consider it.

> + // This is especially useful for making calls to the JRT in places in which this haven't been done before;
> haven't --> hasn't
> 

Yikes.

> methodHandles_ppc.cpp
> 
> As described below, I think it would be nice if you name
> + Register temp1 = R30;
> R30_temp1.
> (This holds also for argbase and param_size, but no need to
> introduce additional changes.)

As I have to edit that file anyway, I'll rename the other locals as well.
Great opportunity to pay off some styling debts...
Using different naming styles in the same function would presumably be much worse.

> Why do you do these register changes here?

`temp1` is used to store the receiver klass and must survive several calls to `MacroAssembler::load_heap_oop` (and thus JRT calls). If it still was a volatile register, we would have to use a higher preservation level; resulting in additional overhead.

As future readers might stumble across this as well, I'll add a comment to make this design choice a little more obvious in that particular case. I also dislike that `tmp1` is still used after the assignment of `temp1` to `temp1_recv_klass`, but I cannot think of a better solution...

The same holds true for all the other register changes.

> Some nice cleanups. The comments with the O5 register etc
> might stem from the sparc port ;)

Although those comments may seem subtle, they actually did lead to some serious confusion!
Nonetheless, quite interesting to find such "artifacts of the past" in the codebase. 

> stubGenerator_ppc.cpp
> 
> + Register tmp1 = R12_tmp, tmp2 = R11_klass;
> I think the register rename is pointless.
> I would just use R12_tmp and R11_scratch1.
> 
> You could move the definition of R11_klass
> below this point, but the list of all regs
> used at the beginning gives a good overview
> of used regs, too.
> 

If I moved R11_klass, I should probably do the same with all the other declarations (if applicable).
So I'll just leave it as it is. 

> templateTable_ppc_64.cpp
> + const Register tmp1 = R11_scratch1,
> + tmp2 = R12_scratch2;
> Why not use R11_scratch1 directly? The name basically
> has the same meaning as tmp. The advantage is that
> the register number is mentioned in the name, which
> makes debugging register problems more easier.
> 
> Look at this artificial example:
> 
> __ load_heap_oop(dst, offset, base, tmp1, tmp2, ...
> looks good.
> __ load_heap_oop(R11_dst, R12_offset, R12_base, tmp1, tmp2,
> makes me suspicious: how can offset and base be the same
> register?

That is a _really_ good point!
I'll change it and keep it in mind for future changes.

Thank you for your review!

**Review by TheRealMDoerr — APPROVED** [review]

**Comment by GoeLin** [comment]
Looks good!

**Review by GoeLin — APPROVED** [review]
Looks good now, thanks!
